File: JavaScript/124-Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    let res = Number.NEGATIVE_INFINITY;
    
    let dfs = (root) => {
        if (!root) return 0;
        let left = Math.max(0, dfs(root.left));
        let right = Math.max(0, dfs(root.right));
        res = Math.max(res, root.val+left+right);
        return root.val + Math.max(left, right);
    }
    
    dfs(root);
    return res;
};

File: C++/124-Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = INT_MIN;
    int dfs(TreeNode* node) {
        if (!node) return 0;
        int leftVal = 0, rightVal = 0;
        leftVal = max(leftVal, dfs(node->left));
        rightVal = max(rightVal, dfs(node->right));
        res = max(res, node->val + leftVal + rightVal);
        
        return max(node->val + leftVal, node->val + rightVal);
    }
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return res;
    }
};